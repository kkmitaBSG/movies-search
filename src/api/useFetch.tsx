import { useEffect, useState } from "react";
import { useSearchParams } from "react-router-dom";
import messages from "../constants/messages";

export interface Items {
  id: string;
  title: string;
  overview?: string;
  original_language?: string;
}

interface UseFetchResponseType {
  status: string;
  results: Items[];
}

export enum LoadingStatus {
  Waiting = "waiting",
  Fetching = "fetching",
  Fetched = "fetched",
}

const useFetch = (searchValue = ""): UseFetchResponseType => {
  const [status, setStatus] = useState<LoadingStatus>(LoadingStatus.Waiting);
  const [results, setData] = useState<Items[]>([]);
  const [, setSearchParams] = useSearchParams();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setStatus(LoadingStatus.Fetching);
        const response = await fetch(
          `https://api.themoviedb.org/3/search/movie?api_key=92b418e837b833be308bbfb1fb2aca1e&query=${searchValue}`
        );

        setSearchParams({ searchParam: searchValue });

        const data = await response.json();
        if (!response.ok) {
          throw response;
        }

        setData(data.results);
        setStatus(LoadingStatus.Fetched);
      } catch {
        console.log(messages.fetchResponse)
      }
    };

    fetchData();
  }, [searchValue, setSearchParams]);

  return { status, results };
};

export default useFetch;
