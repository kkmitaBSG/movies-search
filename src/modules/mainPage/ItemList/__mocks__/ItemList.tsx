import { ReactElement } from "react";

const MockedComponent = (): ReactElement => <>ItemListMockedComponent</>;

export default MockedComponent;
