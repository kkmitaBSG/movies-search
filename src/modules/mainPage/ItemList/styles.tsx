import styled from "styled-components";
import { DebounceInput } from "react-debounce-input";

export const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const Loader = styled.div`
  width: 100%;
  margin: 50px;
  text-align: center;
`;

export const InputText = styled.div``;

export const Input = styled(DebounceInput)`
  width: 100%;
  height: 30px;
  font-size: 20px;
  margin: 10px 0 20px;
  border-radius: 10px;
  padding-left: 10px;
`;
