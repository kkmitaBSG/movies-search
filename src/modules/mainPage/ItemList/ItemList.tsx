import { useState } from "react";
import { useSearchParams } from "react-router-dom";
import useFetch, { LoadingStatus } from "../../../api/useFetch";
import Item from "../Item/Item";
import messages from "../../../constants/messages";
import * as S from "./styles";

const defaultValue = "Rick";

const ItemList = () => {
  const [searchParams] = useSearchParams();
  const [searchValue, setSearchValue] = useState(
    searchParams.get("searchParam") || defaultValue
  );

  const { status, results } = useFetch(searchValue);

  const items = results?.map((item) => <Item key={item.id} {...item} />);

  const changeInput = (ev: any) => {
    // there can not be empty value in req parameter
    if (ev.target.value === "") {
      return setSearchValue(defaultValue);
    }
    setSearchValue(ev.target.value);
  };

  return (
    <>
      <S.InputText>{messages.searchValue} </S.InputText>

      <S.Input
        placeholder={searchValue}
        minLength={1}
        debounceTimeout={500}
        onChange={changeInput}
      />

      <S.Wrapper>
        {status === LoadingStatus.Fetching ? (
          <S.Loader>
            Loading...
          </S.Loader>
        ) : (
          items
        )}
      </S.Wrapper>
    </>
  );
};

export default ItemList;
