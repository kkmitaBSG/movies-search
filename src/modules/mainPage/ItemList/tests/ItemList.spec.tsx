import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import ItemList from "../ItemList";
import useFetch, { Items, LoadingStatus } from "../../../../api/useFetch";

jest.mock("../../item/Item");
jest.mock("../../../../api/useFetch");

const mockData: Items[] = [
  {
    id: "1",
    title: "item1Name",
  },
  {
    id: "2",
    title: "item2Name",
  },
  {
    id: "3",
    title: "item3Name",
  },
];

describe("ItemList", () => {
  test("when data is not fetched yet", () => {
    (useFetch as jest.Mock).mockReturnValue({
      status: LoadingStatus.Fetching,
      data: [],
    });

    const component = renderer.create(
      <MemoryRouter initialEntries={["/"]}>
        <ItemList />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test("when data is empty", () => {
    (useFetch as jest.Mock).mockReturnValue({
      status: LoadingStatus.Fetched,
      data: [],
    });

    const component = renderer.create(
      <MemoryRouter initialEntries={["/"]}>
        <ItemList />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test("when data is filled", () => {
    (useFetch as jest.Mock).mockReturnValue({
      status: LoadingStatus.Fetched,
      data: mockData,
    });

    const component = renderer.create(
      <MemoryRouter initialEntries={["/"]}>
        <ItemList />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
