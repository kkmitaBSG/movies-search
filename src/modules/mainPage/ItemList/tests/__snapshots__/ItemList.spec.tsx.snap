// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList when data is empty 1`] = `
Array [
  <div
    className="sc-dkrFOg"
  >
    Search title:
     
  </div>,
  <input
    className="sc-hLBbgP feKHzH"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Rick"
    type="text"
    value=""
  />,
  <div
    className="sc-bcXHqe kbZGuW"
  />,
]
`;

exports[`ItemList when data is filled 1`] = `
Array [
  <div
    className="sc-dkrFOg"
  >
    Search title:
     
  </div>,
  <input
    className="sc-hLBbgP feKHzH"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Rick"
    type="text"
    value=""
  />,
  <div
    className="sc-bcXHqe kbZGuW"
  />,
]
`;

exports[`ItemList when data is not fetched yet 1`] = `
Array [
  <div
    className="sc-dkrFOg"
  >
    Search title:
     
  </div>,
  <input
    className="sc-hLBbgP feKHzH"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Rick"
    type="text"
    value=""
  />,
  <div
    className="sc-bcXHqe kbZGuW"
  >
    <div
      className="sc-gswNZR fzCsOl"
    >
      Loading...
    </div>
  </div>,
]
`;
